lang/csharp/src/apache/main/POCO/POCOWriter.cs

```diff
@@ -0,0 +1,250 @@
+        }
+
+        /// <summary>
+        /// Serialized an array. The default implementation calls EnsureArrayObject() to ascertain that the
+        /// given value is an array. It then calls GetArrayLength() and GetArrayElement()
+        /// to access the members of the array and then serialize them.
+        /// </summary>
+        /// <param name="schema">The ArraySchema for serialization</param>
+        /// <param name="value">The value being serialized</param>
+        /// <param name="encoder">The encoder for serialization</param>
+        protected override void WriteArray(ArraySchema schema, object value, Encoder encoder)
+        {
+            var arr = value as System.Collections.IList;
+            if (arr == null)
+                throw new AvroTypeException("Array does not implement IList");
+
+            long l = arr.Count;
+            encoder.WriteArrayStart();
+            encoder.SetItemCount(l);
+            for (int i = 0; i < l; i++)
+            {
+                encoder.StartItem();
+                Write(schema.ItemSchema, arr[i], encoder);
+            }
+
+            encoder.WriteArrayEnd();
+        }
+
+        /// <summary>
+        /// Writes the given map into the given encoder.
+        /// </summary>
+        /// <param name="schema">writer schema</param>
+        /// <param name="value">map to write</param>
+        /// <param name="encoder">encoder to write to</param>
+        protected override void WriteMap(MapSchema schema, object value, Encoder encoder)
+        {
+            if (value == null)
+            {
+                throw new AvroTypeException("Map is null - use a union for nullable types");
+            }
+
+            var map = value as System.Collections.IDictionary;
+            if (map == null)
+            {
+                throw new AvroTypeException("Map does not implement IDictionary");
+            }
+
+            encoder.WriteArrayStart();
+            encoder.SetItemCount(map.Count);
+            foreach (System.Collections.DictionaryEntry de in map)
+            {
+                encoder.StartItem();
+                encoder.WriteString(de.Key as string);
+                Write(schema.ValueSchema, de.Value, encoder);
+            }
+
+            encoder.WriteMapEnd();
+        }
+
+        /// <summary>
+        /// Resolves the given value against the given UnionSchema and serializes the object against
+        /// the resolved schema member. The default implementation of this method uses
+        /// ResolveUnion to find the member schema within the UnionSchema.
+        /// </summary>
+        /// <param name="us">The UnionSchema to resolve against</param>
+        /// <param name="value">The value to be serialized</param>
+        /// <param name="encoder">The encoder for serialization</param>
+        protected override void WriteUnion(UnionSchema us, object value, Encoder encoder)
+        {
+            for (int i = 0; i < us.Count; i++)
```

Comment:
`WriteUnion` appears identical to the base implementation. Do we really need to override it?